### ISSUE TITLE
Expose setsockopt in public API

### ISSUE THREAD

**Issue by standy66** [agronholm/anyio#39]
This would be useful for setting up TCP keepalive and TCP_NODELAY flags. Right now the only option is to assume every SocketStream is `anyio._networking.SocketStream` and obtain raw socket via `_socket._raw_socket`

**Comment by agronholm** [comment]
Are you intent on disabling `TCP_NODELAY` for your sockets?

**Comment by standy66** [comment]
I want to be able to execute code similar to the following:
```
socket.setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)
if hasattr(socket, "TCP_KEEPIDLE"):
    socket.setsockopt(socket.IPPROTO_TCP, socket.TCP_KEEPIDLE, 300)
elif is_darwin():
    # Darwin specific option
    # Value extracted from XNU source code
    # https://github.com/apple/darwin-xnu/blob/a449c6a3b8014d9406c2ddbdc81795da24aa7443/bsd/netinet/tcp.h#L207
    TCP_KEEPALIVE = 16
    socket.setsockopt(socket.IPPROTO_TCP, TCP_KEEPALIVE, 300)
socket.setsockopt(socket.IPPROTO_TCP, socket.TCP_KEEPINTVL, 30)
socket.setsockopt(socket.IPPROTO_TCP, socket.TCP_KEEPCNT, 5)
socket.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
```

**Comment by agronholm** [comment]
Yes, I get the use case for `setsockopt()`. I was merely wondering whether or not you knew that `TCP_NODELAY` is set for socket streams by default on anyio.

**Comment by agronholm** [comment]
I implemented `getsockopt()` and `setsockopt()` in `SocketStream` but now I keep wondering if I should expose the raw socket instead. After all, the socket object has a huge amount of low level functionality but I'm uncertain whether it should be exposed or not.

**Comment by standy66** [comment]
Did you test it?
I'm getting `TypeError: setsockopt() takes exactly 3 arguments (4 given)` when calling `socket_stream.setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)`

`socket.SocketType.setsockopt` is a weird built-in method, it does not seem to work when you manually pass `optlen=None`

**Comment by agronholm** [comment]
It seems that I only tested `getsockopt()` but forgot about `setsockopt()`. I will fix this tomorrow.

**Comment by agronholm** [comment]
Fixed in b3832756c1705db93d42c6307ee6749689651f90.
